### ISSUE TITLE
Cookbooks for 'virtual computing lab' scenarios

### ISSUE THREAD

**Issue by aronr** [dlab-berkeley/cloud-computing-working-group#11]
Per a 2015-10-29 discussion with Harrison Dekker, it'd be great to have a cookbook for cloud provisioning of what amounts to a "virtual computing lab" - either a multi-user server or a set of virtual machine instances and other facilities that can be spun up on demand for a training, workshop, or short course.

The use case that Dekker came up with was for setting up R Studio Server and/or a SQL database for trainings. My own recent use case was an inquiry to Research IT by a student teaching a URAP course, who needed to provide students with Chinese-language web scraping software.

My understanding is that folks like @aculich, @ryanlovett, @paciorek, et al. may have many pieces of this they've developed for Stat courses. If we can put together some generic, vanilla templates, scripts and/or instructions for doing this, that'd be of significant benefit.

**Comment by aronr** [comment]
One potential scenario might be a way for students to temporarily spin up and take down virtual hosts for doing homework and practice after workshop/course hours, if those aren't continually running. (The homework scenario applies to the URAP use case above.)

**Comment by aronr** [comment]
There's lots of potential functionality that could be covered in this cookbook. Perhaps those could be separated/isolated and tackled one-by-one?

Examples of those - some of these perhaps already addressed by existing work, as per the previous comment - include:
- Placing constraints on student use of cloud resources, so accidental overages, security breaches, won't incur significant charges.
- Delegating some aspects of administration, such as user account setup and maintenance, to others involved with the event/course. (Example: a faculty member delegating administrative tasks to their GSIs.)

**Comment by paciorek** [comment]
Yes, I agree there are bits and pieces here and there. I think much of what
you mention could be based around BCE with some additional provisioning
steps, or scripts that modify a running VM.  One can use Python's boto
package to programmatically start up instances and to set up multiple IAM
users. Aaron has already provided code for the IAM part of things and I
have some template policy files that limit what individual IAM users on an
account can do. I also have some Boto code that will terminate instances
that run more than X hours for the X of your choice.

The downside to all this is having to deal with either paying Amazon or
getting credits from Amazon. The latter is feasible but their model for
providing credits for groups seems to still be evolving. So it may be hard
to have a cookbook yet with regard to the payment part of it.
